### DISCUSSION TITLE
Python enum not mapped to PostgreSQL enum in schema

### DISCUSSION THREAD

**Discussion by MauroPfister** [sqlalchemy/sqlalchemy#10583] [Usage Questions]
When trying to use a Python enum together with a declarative mapping I run into problems when that enum belongs to a (non-default) schema in a PostgreSQL database. For some reason SQLalchemy does not seem to look up the enum in the correct schema.

**Example**

```Python
import sqlalchemy
from sqlalchemy import MetaData
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from enum import Enum

class TestStatus(Enum):
    PASS = "PASS"
    FAIL = "FAIL"
    SKIP = "SKIP"

metadata_obj = MetaData(schema="example")

class Base(DeclarativeBase):
    metadata = metadata_obj

class Test(Base):
    __tablename__ = "test"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    status: Mapped[TestStatus]
```
When trying to commit instances of the class `Test` to the database I get the following error even though the PostgreSQL enum `
teststatus` exists in the database in the schema `example`:
```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedObject) type "teststatus" does not exist
```

**Temporary solution**
I managed to get things working by replacing the line 
```Python
status: Mapped[TestStatus]
``` 
with

```Python
status: Mapped[TestStatus] = mapped_column(sqlalchemy.Enum(TestStatus, metadata=metadata_obj))
```
However, I'm a bit confused because I thought by specifying a metadata object with the correct schema in the declarative base class, SQLalchemy would look up the enum in the corresponding schema. Is this a wrong assumption and if so can anyone point me to the correct way of handling this?

**Comment by zzzeek** [comment]
you want to use the [inherit_schema](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum.params.inherit_schema) option here.    if that's not working let me know

**Reply by CaselIT** [reply]
it seems that the issue is that the automatically generated enum generated by the type map is not associated with the metadata. probably it should

**Reply by zzzeek** [reply]
maybe.  in any case this default type map enum can be configured to include inherit_schema using the forms documented at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#altering-the-configuration-of-the-default-enum

**Reply by CaselIT** [reply]
interesting, I guess we could decide to go either way. The current behavior is consistent with using an enum type without setting the metadata:
```py
import sqlalchemy as sa
from sqlalchemy import MetaData
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from enum import Enum


class e1(Enum):
    a = 1
    b = 2


class e2(Enum):
    a = 1
    b = 2


class e3(Enum):
    a = 1
    b = 2


class e4(Enum):
    a = 1
    b = 2


class Base(DeclarativeBase):
    metadata = MetaData(schema="test_schema")


class Test(Base):
    __tablename__ = "test"
    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[e1]


class Test2(Base):
    __tablename__ = "test2"
    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[e2] = mapped_column(sa.Enum(e2))


class Test3(Base):
    __tablename__ = "test3"
    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[e3] = mapped_column(sa.Enum(e3, metadata=Base.metadata))


t4 = sa.Table(
    "test4",
    Base.metadata,
    sa.Column("id", sa.Integer),
    sa.Column("status", sa.Enum(e4)),
)


e = sa.create_engine("postgresql://scott:tiger@127.0.0.1:5432/test", echo=True)
with e.connect() as c:
    Base.metadata.create_all(c)
```
prints
```
2023-11-05 12:36:49,985 INFO sqlalchemy.engine.Engine CREATE TYPE e1 AS ENUM ('a', 'b')
2023-11-05 12:36:49,985 INFO sqlalchemy.engine.Engine [no key 0.00009s] {}
...
2023-11-05 12:36:49,990 INFO sqlalchemy.engine.Engine CREATE TYPE e2 AS ENUM ('a', 'b')
2023-11-05 12:36:49,990 INFO sqlalchemy.engine.Engine [no key 0.00011s] {}
...
2023-11-05 12:36:49,991 INFO sqlalchemy.engine.Engine CREATE TYPE test_schema.e3 AS ENUM ('a', 'b')
2023-11-05 12:36:49,991 INFO sqlalchemy.engine.Engine [no key 0.00009s] {}
...
2023-11-05 12:36:49,992 INFO sqlalchemy.engine.Engine CREATE TYPE e4 AS ENUM ('a', 'b')
2023-11-05 12:36:49,992 INFO sqlalchemy.engine.Engine [no key 0.00010s] {}
2023-11-05 12:36:49,993 INFO sqlalchemy.engine.Engine
CREATE TABLE test_schema.test (
        id SERIAL NOT NULL,
        status e1 NOT NULL,
        PRIMARY KEY (id)
)
2023-11-05 12:36:49,993 INFO sqlalchemy.engine.Engine [no key 0.00015s] {}
2023-11-05 12:36:50,002 INFO sqlalchemy.engine.Engine
CREATE TABLE test_schema.test2 (
        id SERIAL NOT NULL,
        status e2 NOT NULL,
        PRIMARY KEY (id)
)
2023-11-05 12:36:50,002 INFO sqlalchemy.engine.Engine [no key 0.00022s] {}
2023-11-05 12:36:50,007 INFO sqlalchemy.engine.Engine
CREATE TABLE test_schema.test3 (
        id SERIAL NOT NULL,
        status test_schema.e3 NOT NULL,
        PRIMARY KEY (id)
)
2023-11-05 12:36:50,007 INFO sqlalchemy.engine.Engine [no key 0.00020s] {}
2023-11-05 12:36:50,016 INFO sqlalchemy.engine.Engine
CREATE TABLE test_schema.test4 (
        id INTEGER,
        status e4
)
2023-11-05 12:36:50,016 INFO sqlalchemy.engine.Engine [no key 0.00021s] {}
2023-11-05 12:36:50,017 INFO sqlalchemy.engine.Engine ROLLBACK
```

only e3 that has been associated with the metadata is in the schema set in the metadata.

> maybe. in any case this default type map enum can be configured to include inherit_schema using the forms documented at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#altering-the-configuration-of-the-default-enum

since this is how enum work in general, we could probably mention it in that part of the docs

**Reply by CaselIT** [reply]
Not sure if this is a bug:
```py
meta = MetaData(schema="test_schema")
class Base(DeclarativeBase):
    metadata = meta
    type_annotation_map = {Enum: sa.Enum(Enum, metadata=meta)}
```
rises `sqlalchemy.exc.CompileError: PostgreSQL ENUM type requires a name.` while this works
```py
meta = MetaData(schema="test_schema")
class Base(DeclarativeBase):
    metadata = meta
    type_annotation_map = {Enum: sa.Enum(Enum, schema=meta.schema)}
```

**Reply by CaselIT** [reply]
> since this is how enum work in general, we could probably mention it in that part of the docs

added it here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4961

**Reply by MauroPfister** [reply]
Thanks a lot for the quick help here! I added a type annotation map to the `Base` class and used the `inherit_schema` option which works perfectly.

```Python
class Base(DeclarativeBase):
    metadata = metadata_obj

    type_annotation_map = {
        enum.Enum: sqlalchemy.Enum(enum.Enum, inherit_schema=True),
    }
```

**Reply by CaselIT** [reply]
look at that. I forgot there was a flag especially for this. I'll update by change above

**Reply by MauroPfister** [reply]
What I still don't fully understand thought, is why this doesn't happen automatically when specifying the metadata in the `Base` class? Is there any reason to assume that the enum wouldn't belong to the same schema?

**Reply by CaselIT** [reply]
that's the default value of `inherit_schema` (``False``). @zzzeek would it make sense to change the default to true for 2.1?

**Reply by zzzeek** [reply]
I think it (so far) makes sense to try and change this default in 2.1.

**Reply by CaselIT** [reply]
ok I'll open an issue so we can reflect on it better

**Reply by rassie** [reply]
Pardon me for hijacking the thread, but what is the correct way to actually create the types in the database? I'm experimenting with moving away from MySQL to PostgreSQL and `Enum`s caught me off-guard. I've got pre-existing Alembic migrations in place, but it seems we used to create `Enum` columns as strings, which somehow still works with fully-annotated model definition. Do I actually need an `Enum` column in the database for the types to be created? Or maybe I can somehow disable type-casting in the queries?

**Reply by zzzeek** [reply]
if you want PostgreSQL enums in Alembic you have to emit the commands yourself, that is, `op.execute("CREATE ENUM ...")` in your migration.  Alembic lacks any kind of direct support for PG enums right now.

**Reply by CaselIT** [reply]
> if you want PostgreSQL enums in Alembic you have to emit the commands yourself, that is, `op.execute("CREATE ENUM ...")` in your migration. Alembic lacks any kind of direct support for PG enums right now.

unless the enum is defined in a `create_table`, since that will create the enum for you

**Reply by rubybui** [reply]
as of now, Ithe way I am doing this is emit the commands myself, any better way to handle this as I did not trust myself nor opus to make change to db
